compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -2194,6 +2188,218 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         Ok((ty, DefKind::AssocTy, assoc_ty_did))
     }
 
+    fn lookup_inherent_assoc_ty(
+        &self,
+        name: Ident,
+        segment: &hir::PathSegment<'_>,
+        adt_did: DefId,
+        self_ty: Ty<'tcx>,
+        block: hir::HirId,
+        span: Span,
+    ) -> Result<Option<(Ty<'tcx>, DefId)>, ErrorGuaranteed> {
+        let tcx = self.tcx();
+
+        let candidates: Vec<_> = tcx
+            .inherent_impls(adt_did)
+            .iter()
+            .filter_map(|&impl_| Some((impl_, self.lookup_assoc_ty_unchecked(name, block, impl_)?)))
+            .collect();
+
+        if candidates.is_empty() {
+            return Ok(None);
+        }
+
+        let param_env = tcx.param_env(block.owner.to_def_id());
+        let cause = ObligationCause::misc(span, block.owner.def_id);
+
+        // In contexts that have no inference context, just make a new one.
+        // We do need a local variable to store it, though.
+        // FIXME(fmease): Shouldn't we do `.ignoring_regions()` like in `method::probe`?
+        let infcx_;
+        let infcx = match self.infcx() {
+            Some(infcx) => infcx,
+            None => {
+                assert!(!self_ty.needs_infer());
+                infcx_ = tcx.infer_ctxt().build();
+                &infcx_
+            }
+        };
+
+        let mut fulfillment_errors = Vec::new();
+        let mut applicable_candidates: Vec<_> = candidates
+            .iter()
+            .filter_map(|&(impl_, (assoc_item, def_scope))| {
+                infcx.probe(|_| {
```

Comment:
If I don't create a snapshot here with `probe`, we only get a single *applicable candidate* in `ambiguity.rs` since the inference variables are “spoiled” in the 2nd iteration. Makes sense.

For some reason, however, the snapshot creation hinders type inference. E.g. in `S<_>::P`, `_` fails to get inferred even if there is only a **single** impl containing the candidate (for example `impl S<()> { type P = (); }`).

The equivalent code with an inherent associated *const* succeeds to infer the parameter. If I get rid of the `probe` call, then `_` does get inferred to `()` in the example given above. I'm pretty sure we want to infer `_` in such a simple case. Correct?

How can I have both things?

Reply:
So, likely need the same thing as candidate assembly then confirmation, where we first assemble possible candidates (in a probe), and then actually confirm that candidate in outside of a probe.